integration_test/third_party_apps_data/applications/dcgm/centos_rhel/install

```diff
@@ -1,54 +1,81 @@
-# less errors on installation; fallback to the runfile method if the package 
-# manager's package is not working or not compitible with the GPU model
-DEVICE_CODE=$(lspci -n | grep -Po '10de:[\w\d]{4}')
-case $DEVICE_CODE in
-    10de:102d)
-        # Install a specific version for NVIDIA Tesla K80, R470 is the last supported version
-        DRIVER_VERSION=470.82.01
-        ;;
-    *)
-        # Installing latest version of NVIDIA CUDA and driver
-        DRIVER_VERSION=535.129.03
-        ;;
-esac
+VERSION_ID=${VERSION_ID%%.*}
 
-echo "Installing NVIDIA Data Center driver $DRIVER_VERSION"
-curl -fSsl -O https://us.download.nvidia.com/tesla/$DRIVER_VERSION/NVIDIA-Linux-x86_64-$DRIVER_VERSION.run
-sudo bash ./NVIDIA-Linux-x86_64-$DRIVER_VERSION.run --silent
+install_driver_from_runfile() {
+    # Ref: https://docs.nvidia.com/datacenter/tesla/tesla-installation-notes/index.html#runfile
+    # This method requires the match kernel-devel package to be installed, and 
+    # the package may be absent from the repo and cause this method to fail
+    sudo yum install -y kernel-devel-$(uname -r) pciutils gcc make wget yum-utils 
+    local DRIVER_VERSION=535.129.03
+    echo "Installing NVIDIA Data Center driver $DRIVER_VERSION"
+    curl -fSsl -O https://us.download.nvidia.com/tesla/$DRIVER_VERSION/NVIDIA-Linux-x86_64-$DRIVER_VERSION.run
+    sudo bash ./NVIDIA-Linux-x86_64-$DRIVER_VERSION.run --silent
 
-# check NVIDIA driver installation succeeded
-nvidia-smi
+    # check NVIDIA driver installation succeeded
+    nvidia-smi
+}
 
-# Install DCGM
-VERSION_ID=${VERSION_ID%%.*}
-sudo yum-config-manager \
-    --add-repo http://developer.download.nvidia.com/compute/cuda/repos/rhel$VERSION_ID/x86_64/cuda-rhel$VERSION_ID.repo
-sudo yum clean expire-cache
-sudo yum install -y datacenter-gpu-manager
-sudo systemctl --now enable nvidia-dcgm
-
-# check DCGM service running and load profiling module
-dcgmi discovery --list
+install_dcgm() {
+    # Ref: https://docs.nvidia.com/datacenter/dcgm/latest/user-guide/getting-started.html#rhel-centos-rocky-linux
+    sudo yum install -y yum-utils
+    sudo yum-config-manager \
+    --add-repo https://developer.download.nvidia.com/compute/cuda/repos/rhel$VERSION_ID/x86_64/cuda-rhel$VERSION_ID.repo
+    sudo yum clean expire-cache
+    sudo yum install -y datacenter-gpu-manager
+    sudo systemctl --now enable nvidia-dcgm 
+
+    # check DCGM service running and load profiling module
+    dcgmi discovery --list
+}
+
+try_install() {
+    for install_method in "$@"
+    do
+        echo "Installing NVIDIA driver with $install_method..."
+        # Run the installation method in a subshell with -e so that the function
+        # exit on error
+        set +e ; ( set -e; $install_method ) ; exit_code=$?
+        set -e
+        if [ $exit_code -eq 0 ]; then
+            echo "NVIDIA driver has been installed successfully with $install_method."
+            return 
+        fi
```

Comment:
Since your main goal here is to ignore the exit code, this could be made much simpler:
```suggestion
        $install_method && {
            echo "NVIDIA driver has been installed successfully with $install_method."
            return
        }
```

Reply:
Here I was also looking to get the behaviour that the function `$install_method` can exit on errors (do not run the remains commands in the function) using `set +e ; ( set -e; func) ;  set -e`. Added comments to better document this; is there a way to simplify the command while keep this?
```
 # Temporarily disable the "-e" option so that failures in the function
 # won't cause this script to exit; also run the installation method in a
 # subshell with -e so that the function itself will exit immediately if 
 # any command encounters error
 ```

Reply:
That's what I was saying — using `&&` will have the effect of temporarily disabling `errexit` for that one command.

Reply:
Sorry here was what I wanted to ask: `&&` would disable `errexit` on calling the function, and evey commands of the function. So with my function
```
install_driver_from_runfile () {
   install kernel_devel
   install driver
}
``` 
`install driver` will still run if `install kernel_devel` fails, when using `install_driver_from_runfile && ...`, and would polluate the logs.
The behaviour I want to have is: disable `errexit` on calling the function, but enable `errexit` on evey commands of the function. 

Another way I can find is to do proper error checking in the function on the commands, so that we don't need to rely on errexit.

Reply:
Ah, you're right. I forgot that bash has this weird behavior that the conditional/boolean context was propagated all the way down into functions and subshells. There's a great explanation at https://lists.gnu.org/archive/html/bug-bash/2012-12/msg00094.html if you wanted to quote it in the comment…

Reply:
Actually, I did find a sort of shorter solution, which uses an explicit `bash` subcommand instead of a subshell. It's not as readable as what I was proposing, so I'll leave it to you to decide whether it's actually worth pursuing. I also added a proposed comment, which you can feel free to omit:
```diff
-        # Run the installation method in a subshell with -e so that the function
-        # exit on error
-        set +e ; ( set -e; $install_method ) ; exit_code=$?
-        set -e
-        if [ $exit_code -eq 0 ]; then
-            echo "NVIDIA driver has been installed successfully with $install_method."
-            return 
-        fi
+        export -f $install_method
+        # Can't use a subshell because of https://lists.gnu.org/archive/html/bug-bash/2012-12/msg00094.html
+        bash -$- -c $install_method && {
+            echo "NVIDIA driver has been installed successfully with $install_method."
+            return
+        }
```
FWIW, this solution is also robust with respect to putting the calling function/shell into the conditional/boolean context.

Reply:
Done! 

Updated to use bash subcommand. Thanks for the detailed suggestion and explanation!